Descend when GPS is lost in auto modes

julianoes opened the pull request:
We shouldn't just drift off when GPS is lost but descend.

This has been tested in SITL using `gpssim stop` but definitely needs outdoor testing. We'll need to come up with a way to fake GPS lost in flight.

@tumbili: thoughts?

@LorenzMeier: comments?

Fixes #4534.

julianoes commented:
@tumbili feedback please.

julianoes commented:
Bump @tumbili, @LorenzMeier please feedback!

A user just reported the same thing in #4674.
We need to merge this, it's better than what we have.

LorenzMeier commented:
Can you rebase and *not* change the GPS check? That will fail arming for anyone wanting to fly manually.

julianoes commented:
@LorenzMeier I iterated more on this and found more things that seem not working at all, for instance the NAV_RCL_ACT param.

I'll do more testing before we can merge it.

julianoes commented:
@LorenzMeier I would say this is now a minimal change.

LorenzMeier commented on src/modules/commander/state_machine_helper.cpp:
```diff
@@ -664,15 +697,17 @@ bool set_nav_state(struct vehicle_status_s *status, struct commander_state_s *in
 
 		/* finished handling commands which have priority, now handle failures */
 		} else if (status_flags->gps_failure) {
-			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LANDGPSFAIL;
```
Where did the land gps fail state go? Should that be removed in the Navigator accordingly? I'm not saying we should, but I'd like to understand the ramifications.

julianoes commented:
I have flown this outside on a Pixfalcon racer, and RTL worked on RC loss, however it didn't auto disarm on landed but went back into RTL.

I couldn't test GPS failure, we'll have to rely on SITL for this, I guess.

I think we can merge this once the RTL-landed-RTL toggle is fixed.

LorenzMeier commented:
Ok - sounds like the toggle should be quickly fixable.

julianoes commented:
It's tricky because we need the `landed` check because we don't want to RTL if we're already landed.

LorenzMeier commented:
Don't think so. How about staying in RTL instead? We can then switch to HOLD / loiter on auto-disarming

julianoes commented:
Hm, we could prevent the navigator from starting a RTL sequence if landed, maybe that's already in there.

julianoes commented:
Yes, it should be:
https://github.com/PX4/Firmware/blob/master/src/modules/navigator/rtl.cpp#L103

julianoes commented:
I cherry-picked 7481632c81a1019997f6b2d7fe61448963ce41cb but it doesn't seem to be enough.

It toggles between landed/in-air somehow, I'll need to dive deeper why that happens:

```
INFO  Landed: no
INFO  throttle: 0.40
INFO  Landed: no
INFO  throttle: 0.40
INFO  Landed: yes
INFO  LANDING DETECTED
INFO  throttle: 0.40
INFO  Landed: no
INFO  TAKEOFF DETECTED
INFO  RTL: on activation
INFO  RTL: land at home
INFO  throttle: 0.40
INFO  Landed: yes
INFO  RTL: completed, landed
INFO  LANDING DETECTED
INFO  throttle: 0.40
INFO  Landed: no
INFO  TAKEOFF DETECTED
INFO  RTL: on activation
INFO  RTL: land at home
INFO  throttle: 0.40
INFO  Landed: yes
INFO  RTL: completed, landed
INFO  LANDING DETECTED
INFO  throttle: 0.40
INFO  Landed: no
INFO  TAKEOFF DETECTED
INFO  RTL: on activation
INFO  RTL: land at home
INFO  throttle: 0.40
INFO  Landed: yes
INFO  RTL: completed, landed
INFO  LANDING DETECTED
INFO  throttle: 0.40
INFO  Landed: no
INFO  TAKEOFF DETECTED
INFO  RTL: on activation
INFO  RTL: land at home
INFO  throttle: 0.40
INFO  Landed: no
INFO  throttle: 0.40
INFO  Landed: no
INFO  throttle: 0.40
INFO  Landed: no
INFO  throttle: 0.40
INFO  Landed: no
INFO  throttle: 0.40
```

julianoes commented:
This is the nominal case when RTL is commanded:

```
INFO  throttle: 0.48
INFO  Landed: no
INFO  throttle: 0.48
INFO  Landed: no
INFO  throttle: 0.48
INFO  Landed: yes
INFO  RTL: completed, landed
INFO  LANDING DETECTED
INFO  throttle: 0.48
INFO  Landed: yes
INFO  throttle: 0.48
INFO  Landed: yes
INFO  throttle: 0.48
INFO  Landed: yes
INFO  throttle: 0.48
INFO  Landed: yes
```

julianoes commented:
@LorenzMeier: I've fixed the problem where it goes back into RTL after having landed on RC loss. It needed @ender-chen's commit, as well as one on top because I had added another switch case for POSCTL that needed the same fix.

I've tested this on SITL, and bench tested it with Pixhawk, and flown a previous version on Pixfalcon earlier, so I think this is safer than what we had before and therefore good to merge.

julianoes commented:
@LorenzMeier, @tumbili Can anyone review or test fly this? I'd like to get it merged as soon as possible.

kd0aij commented:
@julianoes Can you suggest a safe way to test this?

julianoes commented:
The way I tested it in SITL is by using QGC joystick input as RC input, so when you close QGC it reports it as RC lost.

For GPS loss, I typed `gpssim stop` in SITL.

In outdoor flights, all I tested was switching off RC and checking if it RTLs. That's not particularly safe but not too crazy either with my small racer. Unfortunately, we don't have an easy switch to kill GPS on the fly yet.

kd0aij commented:
Thanks. Now that you mention it, I have one 250 with GPS, so I could do the same with it. Or hack in an RC command to simulate both RC and GPS loss.

kd0aij commented:
Does RTL actually mean return and land? My copter descended to perhaps 10m AGL, then went back up again.

whyscience commented:
@kd0aij you should set RTL_LAND_DELAY to not -1

julianoes commented:
Did any testing happen here? If not, I'm strongly suggesting to get it in.

julianoes commented:
This has been flown on a Pixfalcon quad again yesterday and everything was as expected. Let's wait for the CI checks after the last rebase before merging.

julianoes commented:
Ah this got merged, thanks :)
